midi_trainable_tokenizers/bpe_tokenizer.py

```diff
@@ -0,0 +1,106 @@
+import json
+from concurrent.futures import ThreadPoolExecutor
+
+import tokenizers
+import pandas as pd
+from datasets import Dataset
+from tokenizers import Regex, Tokenizer, models, trainers, pre_tokenizers
+
+from midi_tokenizers.midi_tokenizer import MidiTokenizer
+from object_generators.base_tokenizer_generator import BaseTokenizerGenerator
+from midi_trainable_tokenizers.trainable_tokenizer import MidiTrainableTokenizer
+
+
+class BpeMidiTokenizer(MidiTrainableTokenizer):
+    def __init__(self, base_tokenizer: MidiTokenizer, bpe_tokenizer: Tokenizer = None, max_vocab_size: int = None):
+        super().__init__()
+        self.base_tokenizer = base_tokenizer
+        self.tokenizer = bpe_tokenizer
+        self.name = "BpeTokenizer"
+        self.max_vocab_size = max_vocab_size
+
+        if self.max_vocab_size is None:
+            self.max_vocab_size = 30000  # default BpeTrainer vocab_size
+
+        if self.tokenizer is None:
+            # Initialize empty tokenizer and a trainer
+            self.tokenizer = Tokenizer(model=models.BPE())
+            self.tokenizer.pre_tokenizer = self.prepare_text_pre_tokenizer()
+            self.tokenizer.model = models.BPE()
+            self.trainer = trainers.BpeTrainer(
+                vocab_size=self.max_vocab_size,
+                max_token_length=512,
+                special_tokens=["<CLS>"],
+            )
+
+        self.vocab = self.tokenizer.get_vocab()
+
+    def prepare_data_for_training(self, file_name: str, train_dataset: Dataset):
+        def process_record(record):
+            notes = pd.DataFrame(record["notes"])
+            tokens = self.base_tokenizer.tokenize(notes=notes)
+            return " ".join(str(token) for token in tokens) + "\n"
+
+        with open(file=file_name, mode="w+") as file, ThreadPoolExecutor() as executor:
+            # Process records concurrently
+            for result in executor.map(process_record, train_dataset):
+                file.write(result)
+
+    def prepare_text_pre_tokenizer(self):
+        # We have to use this - we cannot load saved tokenizer otherwise
+        byte_level_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=False, use_regex=False)
+
+        # split the tokens into groups before training (concatenate only time tokens and velocity+note_on tokens)
```

Comment:
Don't just write what the code is doing in a comment, a good comment should explain what and **why** is the code doing.